### PR TITLE
[extension/cgroupruntime]: Initial implementation

### PR DESCRIPTION
**Pull request by rogercoll**
**Description:** <Describe what has changed.>

This PR adds the initial implementation of a new component to dynamically set the values of `GOMEMLIMIT` and `GOMAXPROCS` used by the Go runtime. Those values are normally manually aligned with the cgroup resource limit to prevent cpu throttling or out of memory scenarios.

 The component would ease the manual steps of configuring these environment variables in K8s deployments (e.g Helm [templates](https://github.com/open-telemetry/opentelemetry-helm-charts/blob/main/charts/opentelemetry-collector/templates/_helpers.tpl#L169)) in addition to have fine-grained values (e.g. 90% of the resource memory limits).

**Link to tracking Issue:** <Issue number if applicable> https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30289

**Testing:** <Describe what testing was performed and which tests were added.> Unit testing for the component has been added (config and extension start/stop). But ideally, an integration test that actually asserts the runtime modifications should be added as well. The extension relies on "github.com/KimMachineGun/automemlimit/memlimit" and "go.uber.org/automaxprocs/maxprocs" packages for the runtime modifications, but they don't provide a way to mock the "cgroups" file system which is the one they read to get the resource quota limits.

- Automemlimit package tests expect to run in a cgroup environment: https://github.com/KimMachineGun/automemlimit/blob/main/memlimit/cgroups_test.go#L18
- Automaxprocs does not expose the cpu quota retrieval https://github.com/uber-go/automaxprocs/blob/master/maxprocs/maxprocs.go#L41

Any suggestion on how to perform this integration tests in the contrib repository? One possibility is to use the https://github.com/containerd/cgroups package to set the quota, but this requires privileged permissions (also in the GHA)

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by rogercoll** [comment]
Manually tested with a modified version of the OpenTelemetry collector that exposes the internal runtime metrics: https://pkg.go.dev/runtime/metrics

Deployed with very limited cgroup resources:

```
        resources:
          requests:
            cpu: "50m"      # Minimal CPU request
            memory: "64Mi"  # Minimal memory request
          limits:
            cpu: "100m"     # Minimal CPU limit
            memory: "128Mi" # Minimal memory limit
```


Resulting runtime metrics:

```
# HELP go_gc_gomemlimit_bytes Go runtime memory limit configured by the user, otherwise math.MaxInt64. This value is set by the GOMEMLIMIT environment variable, and the runtime/debug.SetMemoryLimit function. Sourced from /gc/gomemlimit:bytes
# TYPE go_gc_gomemlimit_bytes gauge
go_gc_gomemlimit_bytes 1.20795955e+08

...
# HELP go_sched_gomaxprocs_threads The current runtime.GOMAXPROCS setting, or the number of operating system threads that can execute user-level Go code simultaneously. Sourced from /sched/gomaxprocs:threads
# TYPE go_sched_gomaxprocs_threads gauge
go_sched_gomaxprocs_threads 1
```

**Comment by mx-psi** [comment]
@rogercoll Yes! Happy to review this whenever you let me know it is ready

**Comment by rogercoll** [comment]
@mx-psi Great! Moved the PR to ready for review, either for this PR or for a follow-up I think it would be great to have a way to create and run the corresponding integration tests. Thanks!

**Comment by rogercoll** [comment]
Extension "info" logs:

```
2024-11-20T09:59:53.206Z        info    extensions/extensions.go:39     Starting extensions...
2024-11-20T09:59:53.206Z        info    extensions/extensions.go:42     Extension is starting...        {"kind": "extension", "name": "cgroupruntime"}
2024-11-20T09:59:53.257Z        debug   cgroupruntimeextension@v0.112.0/factory.go:46   maxprocs: Updating GOMAXPROCS=[1]: using minimum allowed GOMAXPROCS     {"kind": "extension", "name": "cgroupruntime"}
2024-11-20T09:59:53.257Z        info    cgroupruntimeextension@v0.112.0/extension.go:50 GOMAXPROCS has been set {"kind": "extension", "name": "cgroupruntime", "GOMAXPROCS": 1}
2024-11-20T09:59:53.258Z        info    cgroupruntimeextension@v0.112.0/extension.go:61 GOMEMLIMIT has been set {"kind": "extension", "name": "cgroupruntime", "GOMEMLIMIT": 120795955}
2024-11-20T09:59:53.258Z        info    extensions/extensions.go:59     Extension started.      {"kind": "extension", "name": "cgroupruntime"}


...
2024-11-20T10:00:27.425Z        info    extensions/extensions.go:66     Stopping extensions...
2024-11-20T10:00:27.425Z        debug   cgroupruntimeextension@v0.112.0/factory.go:46   maxprocs: Resetting GOMAXPROCS to [20] {"kind": "extension", "name": "cgroupruntime"}
2024-11-20T10:00:27.425Z        info    service@v0.112.0/service.go:280 Shutdown complete.
```

File: extension/cgroupruntimeextension/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+| [Code Owners](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#becoming-a-code-owner)    | [@mx-psi](https://www.github.com/mx-psi), [@rogercoll](https://www.github.com/rogercoll) |
+
+[development]: https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/component-stability.md#development
+[contrib]: https://github.com/open-telemetry/opentelemetry-collector-releases/tree/main/distributions/otelcol-contrib
+<!-- end autogenerated section -->
+
+## Overview
+
+The OpenTelemetry CGroup Auto-Config Extension is designed to optimize Go runtime performance in containerized environments by automatically configuring GOMAXPROCS and GOMEMLIMIT based on the Linux cgroup filesystem. This extension leverages [automaxprocs](https://github.com/uber-go/automaxprocs) and [automemlimit](https://github.com/KimMachineGun/automemlimit) packages to dynamically adjust Go runtime variables, ensuring efficient resource usage aligned with container limits.
```
Comment (mx-psi) [inline extension/cgroupruntimeextension/README.md:18]: nit: I would use either "Cgroup" or "cgroup"
```suggestion
The OpenTelemetry Cgroup Auto-Config Extension is designed to optimize Go runtime performance in containerized environments by automatically configuring GOMAXPROCS and GOMEMLIMIT based on the Linux cgroup filesystem. This extension leverages [automaxprocs](https://github.com/uber-go/automaxprocs) and [automemlimit](https://github.com/KimMachineGun/automemlimit) packages to dynamically adjust Go runtime variables, ensuring efficient resource usage aligned with container limits.
```

File: extension/cgroupruntimeextension/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+## Examples
+
+```yaml
+extension:
+    # processor name: cgroupruntime
+    cgroupruntime:
+      gomaxprocs:
+        enable: true
+      gomemlimit:
+        enable: true
```
Comment (mx-psi) [inline extension/cgroupruntimeextension/README.md:40]:
```suggestion
      gomaxprocs:
        enabled: true
      gomemlimit:
        enabled: true
```

File: extension/cgroupruntimeextension/config.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+type GoMemLimitConfig struct {
+	Enabled bool    `mapstructure:"enabled"`
+	Ratio   float64 `mapstructure:"ratio"`
+}
+
+// Validate checks if the extension configuration is valid
+func (cfg *Config) Validate() error {
+	if cfg.GoMemLimit.Ratio < 0 || cfg.GoMemLimit.Ratio > 1 {
+		return errors.New("gomemlimit ratio must be between 0 and 1")
```
Comment (mx-psi) [inline extension/cgroupruntimeextension/config.go:25]: Should we also exclude `0` and `1`?

File: extension/cgroupruntimeextension/testdata/config.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+cgroupruntime:
+cgroupruntime/invalid_ratio:
+  gomemlimit:
+    ratio: 1.1
+cgroupruntime/invalid_ratio_type:
+  gomemlimit:
+    ratio: "not_valid"
```
Comment (mx-psi) [inline extension/cgroupruntimeextension/testdata/config.yaml:7]: This will only test that the type in the config is correct but the error will actually come from `confmap`, not from this extension. Maybe it's more interesting to test for negative values?

File: extension/cgroupruntimeextension/config.go (hunk 3 above)
Comment (rogercoll) [reply extension/cgroupruntimeextension/config.go]: Good point, I think we should only exclude the `0` value because it is equivalent to disabling this setting and the automemlimit package will return an error: https://github.com/KimMachineGun/automemlimit/blob/main/memlimit/provider.go#L20

The `1` value might be kept because it would mean setting the GOMEMLIMIT to the defined cgroup memory limit. 

Changed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35472/commits/6abcf495a3d6d457cfb9670f1d459fdcdf815fad

File: extension/cgroupruntimeextension/testdata/config.yaml (hunk 4 above)
Comment (rogercoll) [reply extension/cgroupruntimeextension/testdata/config.yaml:13]: Absolutely, extended the tests cases with a negative and a zero value: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35472/commits/6abcf495a3d6d457cfb9670f1d459fdcdf815fad

**Review by mx-psi — APPROVED** [review]

**Comment by mx-psi** [comment]
I think this can be merged without the integration tests, but some sort of tests should be a requirement for inclusion in the contrib distro. @rogercoll Could you file a separate issue for this so we can track it?

**Comment by rogercoll** [comment]
> I think this can be merged without the integration tests, but some sort of tests should be a requirement for inclusion in the contrib distro. @rogercoll Could you file a separate issue for this so we can track it?

Make sense to me, tracking issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/36545
